### DISCUSSION TITLE
Persisting of the Cart object

### DISCUSSION THREAD

**Discussion by runelaenen** [shopware/platform#3153] [RFC / Proposal]
**Effort**: middle/high

### Description: 
Currently the Cart object is serialized & deserialized in its entirety to the database. This works great, but has 2 large drawbacks:
1) Carts cannot be queried, filtered, ... They always have to be loaded into memory, even to check 1 line item for example
2) Carts with lots of products will throw errors like `1153 Got a packet bigger than 'max_allowed_packet' bytes" ` because we try to write too much info to the database.

### Benefits
- Less errors "1153"

### Break strategy
No need for breaking changes. Only the underlying way of saving the cart should be refactored.

**Comment by JoshuaBehrens** [comment]
As they are serialized I like to quote myself from https://github.com/shopware/proposal/issues/18#issuecomment-1036090275

> PHP serialization is also an issue regarding compatibility BC breaks for internal variables. Usage of serialize makes private variables "public API" and thus break everything that has been serialized/on the queue once.

So everything that is custom in a cart can break when the class is not available anymore. Currently this is bypassed by using only the lineitem payload. This is a good way IMHO but it is still prone to break because someone used all the public API they had in that moment (extensions on structs/everything).

**Comment by rattananen** [comment]
We can use JSON instead of PHP serialized string and use [symfony serializer](https://symfony.com/doc/current/serializer.html) to  serialize/deserialize between JSON and object. 
This way should have no break change and we go benefit that MySQL was supported query in JSON.
Not only Cart object. I think all field that save PHP serialized string should change to JSON instead of BLOB.

**Comment by JoshuaBehrens** [comment]
Due to bugs in MySQL/MariaDB handling JSON we should not tell the database, that it just received JSON :D Beside that I support the comment from @rattananen

**Comment by AndreasA** [comment]
Using JSON instead of serialization is probably a good idea in general as it avoid issues if an object has been changed or does not exist anymore (e.g. cart extensions). However, that is also an issue  because a cart can have extensions that need to be serialized and unserialized and doing so with plain JSON doesn't tell anyone which object the extension is based on, 
so some additional information would be necessary to determine the object to use (does not need to be stored in the DB necessarily could be stored externally and in the worst case fallback to ArrayStruct).

**Comment by rattananen** [comment]
@AndreasA We can leave data in Cart object payload as array.
if extension want to use data in payload they could work with array
or  use [symfony serializer](https://symfony.com/doc/current/components/serializer.html#recursive-denormalization-and-type-safety) denormalize Array to their Object. 
I use this method to work with cart payload in my plugin.

I think save PHP serialized string to database is bad idea because we must use only PHP to deserialized that data.
It's complicate if we want to use other language to handle data in database.

**Comment by AndreasA** [comment]
@rattananen I agree. I just wanted to mention that when this is changed, whoever creates the PR has to makes sure that such things can be stored correctly. Personally I think using symfony serializer with an event to pass on config options would be a good idea or only support array struct for extensions in the cart object and serialize those correctly?

**Comment by JoshuaBehrens** [comment]
>  It's complicate if we want to use other language to handle data in database.

@rattananen do you want to encourage to change data by bypassing the API for it?

@shopware Do we have an API for carts?

**Comment by JanPietrzyk** [comment]
@JoshuaBehrens Not a DAL created one. The cart is not even a DAL Entity. See `\Shopware\Core\Checkout\Cart\CartPersister`, all API cart operations are process driven.

**Comment by JoshuaBehrens** [comment]
@JanPietrzyk  Snap. I had hopes that this changed somewhere in the past 😬  but I understand. This way business rules can be applied very reliably.

**Comment by rattananen** [comment]
@JoshuaBehrens No, I don't but people have their own way to program. 
We don't need to stick with only PHP.  
We should allow them if they can create better tool by other language, and other language has only function for handle JSON but no function for handle PHP serialized string.

**Comment by OliverSkroblin** [comment]
Hey together,
I'm glad that here is so lively discussion about such a "simple" topic 😃 
As background why we have used php serialized at that time: clearly performance 😉 

But I also think that json will be the right choice here. The hint on the `max_allowed_packet` is great, thanks for that, and will definitely find application with JSON. 

We have also already thought about normalizing the values completely and providing an (entity) API for this. 
Here only the issue of performance is important. The reading of the cart must not become much slower due to API support.

**Comment by rattananen** [comment]
@OliverSkroblin About performance, I think memory consumption is should be priority over performance because 
when we use more memory this mean we need to use more time read/write memory. 
And I think EntityCollection cause this problem. now when we search by use repository->search, DAL with create all Entities and dump to memory.  

example 
Person Entity has property name is string 255. 
unicode use 4 bytes per character. 
and repository->search got 100K result. 
so 4 * 255 * 100K = 102MB allocate to memory for this result is 20% of Shopware6 PHP memory_limit requirement.  

This should be one reason to make Shopware slow.

so I think DAL should provide API for operate entity one by one. maybe create LazyEntityCollection or EntitySearchResult->fetch is ok.

**Comment by JoshuaBehrens** [comment]
>  and repository->search got 100K result.

@rattananen When do you want to have 100K in memory? There is e.g. the RepositoryIterator which automatically paginates for you. Maybe this helps out to use the API in a more performant way.

>  Person Entity has property name is string 255.

@rattananen Person entity will also have hex primary key. So you can add 32 characters there as well 😬  I know this is not better but it is always memory you consume as well to it. So we might even be at 115MB?

**Comment by rattananen** [comment]
@JoshuaBehrens  Thank you for RepositoryIterator.  
But RepositoryIterator is also create another problem. it's split one query to multiple queries.
If I have complex criteria that create query has execute time for 0.1sec
I set to 1000 result per fetch to avoid memory problem and I have 100K results.
mean RepositoryIterator need query 100K/1000 = 100 time
so I need time for query 0.1 * 100 = 10 sec.  (It should lower for reduction from select process  in Mysql but it's little compare with join table process)

actually I can optimized to 5000  result per fetch then I will got 2 sec and use 20.4KB
This better result but require thinking so much.
if DAL provide query one time + fetch for one by one should make everything easier.

My example is just simple case. real cases are worse than mine.

**Comment by JoshuaBehrens** [comment]
I agree that working efficiently with DAL is neither easy nor clear in the outcome of e.g. SQL joins.

**Comment by AndreasA** [comment]
Regarding DAL and SQL joins I agree. Well actually DAL is quite limited in various other features as well (e.g. ordering by functions instead of just fields etc. - though it works fine in many scenarios), but I am not really sure what exactly this has to do with persisting the cart object. Might need another another proposal for this 😄

**Comment by mstegmeyer** [comment]
Hey everyone,
this discussion has become a bite stale, sorry about that.
Since Shopware 6.6, we're using compression to reduce the size of the serialized object and we haven't mentioned it here, shame on us.
Does this make the thought about moving to JSON serialization or DAL persisting redundant? If so, I would close this discussion.
